### PR TITLE
feat(curriculum): add bookmark manager lab

### PR DESCRIPTION
**Pull request by Dario-DC**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

File: curriculum/challenges/english/25-front-end-development/lab-bookmark-manager-app/66def5467aee701733aaf8cc.md [hunk 1]
```diff
@@ -0,0 +1,781 @@
...
+
+## --after-user-code--
+
+```js
+const viewCategoryButtonTest = document.getElementById("view-category-button");
+const closeListButtonTest = document.getElementById("close-list-button");
+const closeFormButtonTest = document.getElementById("close-form-button");
+const addBookMarkButtonTest = document.getElementById("add-bookmark-button");
+const addBookmarkButtonFormTest = document.getElementById("add-bookmark-button-form");
+const deleteBookmarkButtonTest = document.getElementById("delete-bookmark-button");
```
Comment (zairahira) [inline curriculum/challenges/english/25-front-end-development/lab-bookmark-manager-app/66def5467aee701733aaf8cc.md:300]:
```suggestion
const deleteBookmarkButton = document.getElementById("delete-bookmark-button");
```

Not sure if this is the actual issue, but the last test was failing with the error: 

```js
// console output
[ReferenceError: deleteBookmarkButton is not defined]
```

File: curriculum/challenges/english/25-front-end-development/lab-bookmark-manager-app/66def5467aee701733aaf8cc.md [hunk 2]
```diff
@@ -0,0 +1,781 @@
...
+1. Once the `bookmarks` key is updated, you should reset the value of `#name` and `#url` to an empty string before running your function to hide the form section and show the main section.
+1. You should have a function named `displayOrHideCategory` that toggles the `hidden` class on `#main-section` and `#bookmark-list-section`.
+1. When you click `#view-category-button`, you should update the inner text of `.category-name` to be the value of the selected option from `#category-dropdown`, modify the inner HTML of `#category-list` according to the user stories below, and call the `displayOrHideCategory` function.
+1. If none of the bookmarks in local storage have the category, you should add a `p` element with the text `No Bookmarks Found` to `#category-list`'s inner HTML.
+1. If one or more bookmarks in local storage have the selected category, you should add a radio button having the `id` and `value` attributes set to the bookmark name for each bookmark in the category.
+Each radio button should have a corresponding label containing an anchor element with the bookmark name and the `href` attribute set to the bookmark URL.
```
Comment (zairahira) [inline curriculum/challenges/english/25-front-end-development/lab-bookmark-manager-app/66def5467aee701733aaf8cc.md:28]:
```suggestion
1. Each radio button should have a corresponding label containing an anchor element with the bookmark name and the `href` attribute set to the bookmark URL.
```

**Review by zairahira — CHANGES_REQUESTED** [review]

File: client/i18n/locales/english/intro.json [hunk 3]
```diff
@@ -2052,7 +2052,10 @@
       "pehm": { "title": "227", "intro": [] },
       "cvsw": { "title": "228", "intro": [] },
       "qjws": { "title": "229", "intro": [] },
```
Comment (zairahira) [inline client/i18n/locales/english/intro.json:2054]: I noticed that now the order is `229` in the sheet.

File: curriculum/challenges/_meta/lab-bookmark-manager-app/meta.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "Build a Bookmark Manager App",
+  "isUpcomingChange": true,
+  "usesMultifileEditor": true,
+  "blockType": "lab",
+  "dashedName": "lab-bookmark-manager-app",
+  "order": 230,
```
Comment (zairahira) [inline curriculum/challenges/_meta/lab-bookmark-manager-app/meta.json:7]: The order should be updated to `229`.

**Review by zairahira — APPROVED** [review]

**Review by moT01 — COMMENTED** [review]
This is looking good @Dario-DC 👍 Very challenging and really made me think. I left a few comments below to try and clarify some of the user stories. Additionally, I couldn't see the bookmark list very well:

<details><summary>image</summary>

Screenshot 2024-09-25 at 9 55 51 PM

</details>

I think we should change the CSS a little so you can see the text of those (visited) links - or maybe add a lighter background to the text or something.

Also, here's the JS code I used to pass it if you want to use it for the solution - still room for improvement, but it's a little more concise:

<details><summary>code</summary>

```
const mainSection = document.getElementById('main-section');
const formSection = document.getElementById('form-section');
const addBookmarkBtn = document.getElementById('add-bookmark-button');
const categoryName = document.querySelector('.category-name');
const categoryDropdown = document.getElementById('category-dropdown');
const closeFormBtn = document.getElementById('close-form-button');
const addBookmarkBtnForm = document.getElementById('add-bookmark-button-form');
const bookmarkName = document.getElementById("name");
const bookmarkURL = document.getElementById("url");
const bookmarkListSection = document.getElementById('bookmark-list-section');
const viewCategoryBtn = document.getElementById('view-category-button');
const categoryList = document.getElementById('category-list');
const closeListBtn = document.getElementById('close-list-button');
const deleteBookmarkBtn = document.getElementById('delete-bookmark-button');

const getBookmarks = () => JSON.parse(localStorage.getItem("bookmarks")) || [];

const displayOrCloseForm = () => {
  mainSection.classList.toggle("hidden");
  formSection.classList.toggle("hidden");
}

addBookmarkBtn.addEventListener('click', () => {
  categoryName.innerText = categoryDropdown.value;
  displayOrCloseForm();
});

closeFormBtn.addEventListener('click', displayOrCloseForm);

addBookmarkBtnForm.addEventListener('click', () => {
  const newBookmark = {
    name: bookmarkName.value,
    category: categoryDropdown.value,
    url: bookmarkURL.value
  };

  if (bookmarkName.value.length && bookmarkURL.value.length) {
    localStorage.setItem("bookmarks", JSON.stringify([...getBookmarks(), newBookmark]))
    bookmarkName.value = '';
    bookmarkURL.value = '';
    displayOrCloseForm();
  }
});

function displayOrHideCategory() {
  mainSection.classList.toggle("hidden");
  bookmarkListSection.classList.toggle("hidden");
}

viewCategoryBtn.addEventListener('click', () => {
  categoryName.innerText = categoryDropdown.value;
  const bookmarksToShow = getBookmarks().filter(bookmark => bookmark.category === categoryDropdown.value)
  fillBookmarkList(bookmarksToShow);
  displayOrHideCategory();
});

closeListBtn.addEventListener('click', displayOrHideCategory);

deleteBookmarkBtn.addEventListener('click', () => {
  const selectedBookmark = document.querySelector('input[type="radio"]:checked');
  const newBookmarks = getBookmarks().filter(bookmark =>
    bookmark.name !== selectedBookmark.value || bookmark.category !== categoryDropdown.value
  );
  localStorage.setItem("bookmarks", JSON.stringify(newBookmarks));
  const bookmarksToShow = getBookmarks().filter(bookmark => bookmark.category === categoryDropdown.value);
  fillBookmarkList(bookmarksToShow);
});

const fillBookmarkList = (bookmarksToShow) => {
  categoryList.innerHTML = '';
  if (bookmarksToShow.length === 0) {
    categoryList.innerHTML = '<p>No Bookmarks Found</p>'
  } else {
    bookmarksToShow.forEach(bookmark => {
      categoryList.innerHTML += `<div>
        <input type="radio" name="bookmarks" id="${bookmark.name}" value="${bookmark.name}">
        <label for="${bookmark.name}">
          <a href="${bookmark.url}">${bookmark.name}</a>
        </label>
      </div>`
    });
  }
}
```

</details>

**Comment by Dario-DC** [comment]
Thanks for your feedback @moT01 Do you think this could work?
image
image

**Review by moT01 — APPROVED** [review]
🎉

**Pull request merged**